File: python/ascii-art-generator.py
#!/usr/bin/env python3
"""
ASCII Art Generator - Generative Art with Characters
Creates random patterns and shapes using ASCII characters.
"""

import random
import math
from datetime import datetime

class ASCIIArtGenerator:
    def __init__(self, width=80, height=40):
        self.width = width
        self.height = height

    def clear_canvas(self):
        return [[' ' for _ in range(self.width)] for _ in range(self.height)]

    def random_shape(self, canvas):
        """Draw random shapes: circles, rectangles, lines"""
        shape_type = random.choice(['circle', 'rect', 'line', 'spiral'])
        chars = random.choice(['@', '#', '*', '+', 'o', 'x', '~', ':'])
        
        cx, cy = random.randint(0, self.width-1), random.randint(0, self.height-1)
        
        if shape_type == 'circle':
            radius = random.randint(3, min(self.width, self.height) // 4)
            for y in range(self.height):
                for x in range(self.width):
                    dist = math.sqrt((x - cx)**2 + (y - cy)**2)
                    if abs(dist - radius) < 1.5:
                        canvas[y][x] = chars
                        
        elif shape_type == 'rect':
            w, h = random.randint(5, 20), random.randint(3, 10)
            x1, y1 = max(0, cx - w//2), max(0, cy - h//2)
            x2, y2 = min(self.width-1, cx + w//2), min(self.height-1, cy + h//2)
            for x in range(x1, x2+1):
                if 0 <= y1 < self.height: canvas[y1][x] = chars
                if 0 <= y2 < self.height: canvas[y2][x] = chars
            for y in range(y1, y2+1):
                if 0 <= x1 < self.width: canvas[y][x1] = chars
                if 0 <= x2 < self.width: canvas[y][x2] = chars
                
        elif shape_type == 'line':
            x2, y2 = random.randint(0, self.width-1), random.randint(0, self.height-1)
            steps = max(abs(x2 - cx), abs(y2 - cy)) or 1
            for i in range(steps + 1):
                t = i / steps
                x, y = int(cx + (x2 - cx) * t), int(cy + (y2 - cy) * t)
                if 0 <= x < self.width and 0 <= y < self.height:
                    canvas[y][x] = chars
                    
        elif shape_type == 'spiral':
            angle = 0
            radius = 1
            max_radius = min(self.width, self.height) // 3
            x, y = cx, cy
            for _ in range(200):
                nx = int(cx + radius * math.cos(angle))
                ny = int(cy + radius * math.sin(angle))
                if 0 <= nx < self.width and 0 <= ny < self.height:
                    canvas[ny][nx] = chars
                angle += 0.3
                radius += 0.1
                if radius > max_radius:
                    break

    def random_pattern(self, canvas):
        """Add random patterns: dots, noise, waves"""
        pattern_type = random.choice(['dots', 'noise', 'waves', 'gradient'])
        
        if pattern_type == 'dots':
            for _ in range(random.randint(20, 100)):
                x, y = random.randint(0, self.width-1), random.randint(0, self.height-1)
                canvas[y][x] = random.choice(['.', "'", '`', '°'])
                
        elif pattern_type == 'noise':
            for y in range(self.height):
                for x in range(self.width):
                    if random.random() < 0.05:
                        canvas[y][x] = random.choice(['·', ':', '·'])
                        
        elif pattern_type == 'waves':
            chars = ['~', '~', '-', '_', '']
            phase = random.uniform(0, math.pi * 2)
            frequency = random.uniform(0.05, 0.2)
            for y in range(self.height):
                for x in range(self.width):
                    offset = int(5 * math.sin(x * frequency + phase))
                    if 0 <= y + offset < self.height and random.random() < 0.3:
                        canvas[y + offset][x] = random.choice(chars[:-1])
                        
        elif pattern_type == 'gradient':
            char = random.choice(['#', '@', '*', '+', ':', '.', ' '])
            gradient = [c for c in '#@*+:. ']
            for y in range(self.height):
                intensity = y / self.height
                char = gradient[min(int(intensity * len(gradient)), len(gradient)-1)]
                for x in range(self.width):
                    if canvas[y][x] == ' ' and random.random() < 0.1:
                        canvas[y][x] = char

    def generate(self, num_shapes=10, num_patterns=5):
        """Generate a complete ASCII art piece"""
        canvas = self.clear_canvas()
        
        # Draw shapes
        for _ in range(num_shapes):
            self.random_shape(canvas)
            
        # Add patterns
        for _ in range(num_patterns):
            self.random_pattern(canvas)
            
        # Convert to string
        result = '\n'.join(''.join(row) for row in canvas)
        return result

def main():
    print("\n" + "="*80)
    print("🎨 ASCII ART GENERATOR")
    print("="*80)
    
    generator = ASCIIArtGenerator(width=80, height=40)
    
    # Generate 3 different art pieces
    for i in range(1, 4):
        print(f"\n--- Variation {i} ---\n")
        art = generator.generate(
            num_shapes=random.randint(5, 15),
            num_patterns=random.randint(2, 6)
        )
        print(art)
        
    print("\n" + "="*80)
    print(f"Generated at: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print("="*80 + "\n")

if __name__ == "__main__":
    main()

File: scripts/ascii-art-generator.js
#!/usr/bin/env node

/**
 * ASCII Art Pattern Generator
 * Creativity meets mathematics - generating beautiful ASCII art patterns
 */

const PATTERNS = {
  mandelbrot: {
    name: 'Mandelbrot Set',
    chars: ' .:-=+*#%@',
    width: 80,
    height: 40,
    generate: (w, h, chars) => {
      const result = [];
      for (let y = 0; y < h; y++) {
        let line = '';
        for (let x = 0; x < w; x++) {
          const cx = (x - w / 2) / (w / 4) - 0.5;
          const cy = (y - h / 2) / (h / 4);
          let zx = 0, zy = 0;
          let iter = 0;
          const maxIter = chars.length - 1;

          while (zx * zx + zy * zy < 4 && iter < maxIter) {
            const temp = zx * zx - zy * zy + cx;
            zy = 2 * zx * zy + cy;
            zx = temp;
            iter++;
          }

          line += chars[iter];
        }
        result.push(line);
      }
      return result;
    }
  },

  ripple: {
    name: 'Stone Ripple',
    chars: ' .•*◊◉●',
    width: 60,
    height: 30,
    generate: (w, h, chars) => {
      const result = [];
      const cx = w / 2, cy = h / 2;
      for (let y = 0; y < h; y++) {
        let line = '';
        for (let x = 0; x < w; x++) {
          const dist = Math.sqrt((x - cx) ** 2 + (y - cy) ** 2);
          const wave = Math.sin(dist / 3 - Date.now() / 1000);
          const charIndex = Math.floor((wave + 1) / 2 * (chars.length - 1));
          line += chars[Math.max(0, Math.min(chars.length - 1, charIndex))];
        }
        result.push(line);
      }
      return result;
    }
  },

  spiral: {
    name: 'Golden Spiral',
    chars: ' ░▒▓█',
    width: 50,
    height: 50,
    generate: (w, h, chars) => {
      const result = [];
      const cx = w / 2, cy = h / 2;
      for (let y = 0; y < h; y++) {
        let line = '';
        for (let x = 0; x < w; x++) {
          const dx = x - cx, dy = y - cy;
          const dist = Math.sqrt(dx * dx + dy * dy);
          const angle = Math.atan2(dy, dx);
          const spiral = Math.sin(angle * 5 + dist / 5);
          const charIndex = Math.floor((spiral + 1) / 2 * (chars.length - 1));
          line += chars[Math.max(0, Math.min(chars.length - 1, charIndex))];
        }
        result.push(line);
      }
      return result;
    }
  },

  waves: {
    name: 'Ocean Waves',
    chars: ' 〜～∼⁓≈',
    width: 80,
    height: 30,
    generate: (w, h, chars) => {
      const result = [];
      for (let y = 0; y < h; y++) {
        let line = '';
        for (let x = 0; x < w; x++) {
          const wave1 = Math.sin(x / 8 + y / 4);
          const wave2 = Math.sin(x / 6 - y / 3);
          const combined = (wave1 + wave2) / 2;
          const charIndex = Math.floor((combined + 1) / 2 * (chars.length - 1));
          line += chars[Math.max(0, Math.min(chars.length - 1, charIndex))];
        }
        result.push(line);
      }
      return result;
    }
  },

  stars: {
    name: 'Night Sky',
    chars: '  .⋆✦⭐✨',
    width: 70,
    height: 35,
    generate: (w, h, chars) => {
      const result = [];
      const seed = Math.random() * 10000;
      for (let y = 0; y < h; y++) {
        let line = '';
        for (let x = 0; x < w; x++) {
          const noise = Math.sin(x * 12.9898 + y * 78.233 + seed) * 43758.5453;
          const normalized = noise - Math.floor(noise);
          const threshold = 1 - Math.pow(normalized, 3);
          const charIndex = Math.floor(threshold * (chars.length - 1));
          line += chars[Math.max(0, Math.min(chars.length - 1, charIndex))];
        }
        result.push(line);
      }
      return result;
    }
  },

  dna: {
    name: 'DNA Double Helix',
    chars: ' ·░▒▓█',
    width: 60,
    height: 50,
    generate: (w, h, chars) => {
      const result = [];
      const cx = w / 2;
      const amplitude = w / 5;
      const frequency = 0.12;

      for (let y = 0; y < h; y++) {
        let line = '';
        for (let x = 0; x < w; x++) {
          const t = y * frequency;
          const strand1Offset = Math.sin(t) * amplitude;
          const strand2Offset = Math.sin(t + Math.PI) * amplitude;
          const strand1X = cx + strand1Offset;
          const strand2X = cx + strand2Offset;

          const dist1 = Math.abs(x - strand1X);
          const dist2 = Math.abs(x - strand2X);

          const baseDist = Math.min(dist1, dist2);
          const connectionDist = Math.abs(dist1 - dist2);

          let intensity = 0;

          if (baseDist < 1.5) {
            intensity = 1 - (baseDist / 1.5);
          } else if (connectionDist < 2 && Math.abs(Math.sin(t * 2)) > 0.3) {
            intensity = 0.5 * (1 - connectionDist / 2);
          }

          const charIndex = Math.floor(intensity * (chars.length - 1));
          line += chars[Math.max(0, Math.min(chars.length - 1, charIndex))];
        }
        result.push(line);
      }
      return result;
    }
  }
};

function generateArt(patternName) {
  const pattern = PATTERNS[patternName];
  if (!pattern) {
    throw new Error(`Unknown pattern: ${patternName}`);
  }

  console.log(`\n${'='.repeat(pattern.width)}`);
  console.log(`  🎨 ${pattern.name}`);
  console.log(`${'='.repeat(pattern.width)}\n`);

  const art = pattern.generate(pattern.width, pattern.height, pattern.chars);
  art.forEach(line => console.log(line));

  console.log(`\n${'='.repeat(pattern.width)}\n`);

  return art.join('\n');
}

// Generate all patterns
function generateGallery() {
  console.log('\n' + '🖼️  ASCII ART GALLERY '.padEnd(80, '=') + '\n');

  Object.keys(PATTERNS).forEach((patternName, index) => {
    console.log(`\n[${index + 1}/${Object.keys(PATTERNS).length}]`);
    generateArt(patternName);
  });
}

// Main execution
if (require.main === module) {
  const args = process.argv.slice(2);

  if (args.length === 0 || args[0] === 'gallery') {
    generateGallery();
  } else if (args[0] === 'list') {
    console.log('\n📋 Available Patterns:\n');
    Object.entries(PATTERNS).forEach(([key, pattern]) => {
      console.log(`  ${key.padEnd(15)} - ${pattern.name}`);
    });
    console.log('');
  } else if (PATTERNS[args[0]]) {
    generateArt(args[0]);
  } else {
    console.log(`\n❌ Unknown pattern: ${args[0]}`);
    console.log('   Run: node ascii-art-generator.js list');
    console.log('   Or:  node ascii-art-generator.js gallery\n');
    process.exit(1);
  }
}

module.exports = { PATTERNS, generateArt };
